s3: Fix precondition failed in CopyObjectPart when src is encrypted

vadmeste opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
CopyObject precondition checks into GetObjectReader
in order to perform SSE-C pre-condition checks using the
last 32 bytes of encrypted ETag rather than the decrypted
ETag

This also necessitates moving precondition checks for
gateways to gateway layer rather than object handler check

## Motivation and Context
Fix issue

## Regression
No

## How Has This Been Tested?
This PR tests this use case https://github.com/minio/minio-go/pull/1078

Otherwise, to test manually, you can upload a large file, e.g. 600 Mb, upload it
encrypted (SSE-C), then copy it in the same bucket with mc.


## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added unit tests to cover my changes. (https://github.com/minio/minio-go/pull/1078)
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

poornas commented on cmd/encryption-v1.go:
```diff
@@ -923,16 +924,18 @@ func getDecryptedETag(headers http.Header, objInfo ObjectInfo, copySource bool)
-		if err != nil {
-			return objInfo.ETag
-		}
+		return objInfo.ETag[len(objInfo.ETag)-32:]
```
@vadmeste, while this will fix the precondition validation error, it will fail in the case of key rotation. If source and dest are the same - only the metadata is replaced and ETag is just copied over ( which is why it was being decrypted if copySource was true). In this case, ETag will be re-encrypted using the last 32 bytes rather than the md5sum, leading to incompatibility with s3 spec if ssec->s3 copy is done.

harshavardhana replied:
We can detect let's say if the destination is being encrypted using SSE-S3 then we can try to get the key out of the request if not this code is fine.

poornas replied:
seems to me ObjectInfo needs to store encrypted md5 separately so that a helper function can get decrypted ETag if encryption key is passed in. The issue is ObjectInfo will only have decrypted ETag after GetObjectReader returns. We need both decrypted ETag (for key rotation) and last 32 bytes ( for pre-condition validation) in case of SSE-C

harshavardhana replied:
yes, but that refactor is not needed if we simply add one check here..

vadmeste replied:
@poornas are you saying that ETag doesn't change when we rotate keys ? I tested with this script but it looks like the ETag is always changed:

```sh
#!/bin/bash

aws s3api put-object \
  --bucket vadmeste-garage --key my-secret-source-object \
  --sse-customer-algorithm AES256 \
  --sse-customer-key MzJieXRlc2xvbmdzZWNyZXRrZXltdXN0cHJvdmlkZWQ= \
  --sse-customer-key-md5 7PpPLAK26ONlVUGOWlusfg== \
  --body ./README.md

aws s3api head-object --bucket vadmeste-garage --key my-secret-source-object \
 --sse-customer-algorithm AES256 \
 --sse-customer-key MzJieXRlc2xvbmdzZWNyZXRrZXltdXN0cHJvdmlkZWQ= \
 --sse-customer-key-md5 7PpPLAK26ONlVUGOWlusfg==

aws s3api copy-object \
    --bucket vadmeste-garage --key my-secret-source-object \
    --copy-source /vadmeste-garage/my-secret-source-object \
    --metadata-directive REPLACE \
    --sse-customer-algorithm AES256 \
    --sse-customer-key "Qi1sHpQnppeJAo5WyM8w/BGJXJmTO/LBg6dcjThY6nM=" \
    --sse-customer-key-md5 "IcpiSupe46fQ0fb5AGQ2RQ==" \
    --copy-source-sse-customer-algorithm AES256 \
    --copy-source-sse-customer-key MzJieXRlc2xvbmdzZWNyZXRrZXltdXN0cHJvdmlkZWQ= \
    --copy-source-sse-customer-key-md5 7PpPLAK26ONlVUGOWlusfg==


aws s3api head-object --bucket vadmeste-garage --key my-secret-source-object \
 --sse-customer-algorithm AES256 \
 --sse-customer-key "Qi1sHpQnppeJAo5WyM8w/BGJXJmTO/LBg6dcjThY6nM=" \
 --sse-customer-key-md5 "IcpiSupe46fQ0fb5AGQ2RQ=="

```

aead replied:
The reason is that the etag encrypted with a different key. So the encrypted etags will differ but the plain etag is the same.

poornas committed:
Fix CopyObjectPart precondition check when source is ssec encrypted

CopyObject precondition checks into GetObjectReader
in order to perform SSE-C pre-condition checks using the
last 32 bytes of encrypted ETag rather than the decrypted
ETag

This also necessitates moving precondition checks for
gateways to gateway layer rather than object handler check

poornas reviewed (commented):
LGTM and tested. 
might want to remove me from official reviewers for this PR.

harshavardhana commented on cmd/object-api-utils.go:
```diff
@@ -486,7 +488,7 @@ func NewGetObjectReader(rs *HTTPRangeSpec, oi ObjectInfo, cleanUpFns ...func())
 		// encrypted bytes. The header parameter is used to
 		// provide encryption parameters.
-		fn = func(inputReader io.Reader, h http.Header, cFns ...func()) (r *GetObjectReader, err error) {
+		fn = func(inputReader io.Reader, h http.Header, pcfn CheckCopyPreconditionFn, cFns ...func()) (r *GetObjectReader, err error) {
```
S3 behavior requires precondition to be validated before validating the range - but this can be fixed in a different PR @vadmeste

harshavardhana reviewed (dismissed):
LGTM & tested

kannappanr commented on cmd/gateway/s3/gateway-s3.go:
```diff
@@ -467,6 +467,9 @@ func (l *s3Objects) CopyObject(ctx context.Context, srcBucket string, srcObject
 	// metadata input is already a trickled down value from interpreting x-amz-metadata-directive at
 	// handler layer. So what we have right now is supposed to be applied on the destination object anyways.
 	// So preserve it by adding "REPLACE" directive to save all the metadata set by CopyObject API.
+	if srcOpts.CheckCopyPrecondFn != nil && srcOpts.CheckCopyPrecondFn(srcInfo, "") {
```
Code should be above the comment.

aead commented on cmd/encryption-v1.go:
```diff
@@ -923,16 +923,18 @@ func getDecryptedETag(headers http.Header, objInfo ObjectInfo, copySource bool)
-		if err != nil {
-			return objInfo.ETag
-		}
+		return objInfo.ETag[len(objInfo.ETag)-32:]
```
Maybe it's just me but I find the comment quite confusing. In case of SSE-C the etag is encrypted with a separate key (which depends on the object encryption key - see https://github.com/minio/minio/blob/master/cmd/crypto/key.go#L148). Now, the code is *only* correct because the object encryption key is not changed. If the object encryption key would be changed or the etag encryption would depend on e.g. the key-encryption key the code would be wrong b/c the etag would change on a copy... 

Maybe something like:
```
	// If source is SSE-C encrypted and this is a SSE-C Copy operation
	// we should return the last 32 hex chars (16 byte) for the ETag, since the client
        // has only ever seen this (random-looking) ETag for that object. This corresponds
        // to the S3 spec. since the ETag is not the MD5 sum of the object plaintext in case of SSE-C.
```

harshavardhana replied:
@poornas does this change work with SSE-C to SSE-S3 - since this returns a different ETag without decrypting how would it work?

aead commented on cmd/object-api-interface.go:
```diff
@@ -26,10 +26,14 @@ import (
 	"github.com/minio/minio/pkg/policy"
 )
 
+// CheckCopyPreconditionFn returns true if copy precondition check failed.
```
Wait, if the function returns true it's an error?! I suggest to return an error (i.e. `PreConditionFailed`) directly...
So code will read like this:
```
if checkCopyPrecondition(object, etag) != nil {
    // error handling
}
```

poornas replied:
@aead, the precondition check tests more than one header - such as ETag and if-modified-since. If it is written as above, error check will need to be redone to write the correct error headers at the handler.

harshavardhana replied:
That cleanup we might have to do later @aead

harshavardhana commented on cmd/encryption-v1.go:
```diff
@@ -923,16 +923,18 @@ func getDecryptedETag(headers http.Header, objInfo ObjectInfo, copySource bool)
-		if err != nil {
-			return objInfo.ETag
-		}
+		return objInfo.ETag[len(objInfo.ETag)-32:]
```
This change can be reverted since now CopyCondition check handles this appropriately.

poornas replied:
this change needs to be reverted

vadmeste committed:
Check preconditions for gateways during CopyObject & CopyObjectPart

harshavardhana reviewed (approved):
LGTM & tested

poornas reviewed (commented):
Tested, LGTM.
